### ISSUE TITLE
Drag-adjustable panels cause darkroom ui lag in certain width configurations

### ISSUE THREAD

**Issue by jpenala** [darktable-org/darktable#3876]
**Describe the bug:**

After trying to setup the darkroom view to my liking with the new drag-adjustable panels, I noticed that with certain panel widths the preview image and image adjustment sliders have serious lag (stutter/delay). On some panel widths everything is smooth as butter, seems to be quite arbitrary behaviour. And if I happen to find a config that works for one image, the same config causes lag when using it for an image with different dimensions/aspect ratio (from different camera). I also ran -d perf and it seems the lag is just ui lag and not affecting the actual processing speed.

**Steps to reproduce the behavior:**

Import image to darkroom, drag both left and right panels to different widths and then try adjusting exposure with different panel sizes.

**Expected behavior:**
No lag, smooth update of the preview image and no slider stutter.

**Screenshots:**
Configurations with and without lag:

No lag
![no_lag](https://user-images.githubusercontent.com/59263827/71488427-1cca1f80-2829-11ea-9311-68888585ef64.png)

Lag
![lag](https://user-images.githubusercontent.com/59263827/71488433-22276a00-2829-11ea-9fe7-d3c0b7d05c41.png)

Lag
![lag2](https://user-images.githubusercontent.com/59263827/71488438-2784b480-2829-11ea-80fe-3767d06a9c43.png)

**Platform:**
 - OS: Fedora Workstation 31
 - Version: darktable-3.1.0~git132
 - OpenCL on/off makes no difference
 - NVIDIA Quadro  w/ NVIDIA Driver 440.44

**Additional context**

 - Can you reproduce with another dt version?

No, 3.0.0. works smoothly with the hardcoded panel values (using panel width 322px) and similar configuration with filmstrip minimized and rating panel hidden.

**Comment by TurboGit** [comment]
In the lag version, can you exit dt and report what is output with:

`$ grep _size ~/.config/darktable/darktablerc`

Thanks.

**Comment by jpenala** [comment]
Two different configs that lag (for a landscape image Nikon D7000):

Narrower panels, more empty space on the sides of the image
```
[fedora-admin@fedora-workstation ~]$ grep _size ~/.config/darktable/darktablerc
darkroom/ui/0/bottom_size=64
darkroom/ui/0/left_size=293
darkroom/ui/0/right_size=256
lighttable/ui/1/bottom_size=64
opencl_size_roundup=16
plugins/darkroom/spots/circle_size=0.01881788857281208
plugins/darkroom/ui/border_size=5
plugins/lighttable/preview/full_size_preload_count=1
ui_last/colorpicker_size=1
```

Wider panels, more empty space above and below image
```
[fedora-admin@fedora-workstation ~]$ grep _size ~/.config/darktable/darktablerc
darkroom/ui/0/bottom_size=64
darkroom/ui/0/left_size=318
darkroom/ui/0/right_size=402
lighttable/ui/1/bottom_size=64
opencl_size_roundup=16
plugins/darkroom/spots/circle_size=0.01881788857281208
plugins/darkroom/ui/border_size=5
plugins/lighttable/preview/full_size_preload_count=1
ui_last/colorpicker_size=1
```

Also between these there are "zones" that don't have lag and others that have it, quite random.

**Comment by junkyardsparkle** [comment]
Interesting...with today's build I noticed lag and high CPU usage just from activating a masking brush and moving the mouse over the image (no clicking or actual mask painting, just positioning movement). I didn't figure out that it was related to the panels (which I did drag-adjust), but here's the dimensions I changed, for whatever it's worth as another data point:

darkroom/ui/0/left_size=302
darkroom/ui/0/right_size=311

**Comment by jpenala** [comment]
Yeah I get that too, just moving the brush over the image.

Here's two comparisons of processor usage while moving tone curve up and down, the top one is the laggy config and bottom one is normal smooth sailing. Wicked single core spiking going on (I disabled OpenCL for these tests):

![no-opencl](https://user-images.githubusercontent.com/59263827/71507568-a14e8980-288d-11ea-89f6-761c5651687a.jpg)

Also might be worth noting: if some panel config causes adjustment slider lag in "fit to screen" mode, the lag might disappear when zooming in. But then panning of the zoomed image stutters like crazy and causes similar cpu spikes. I also run tests with high performance cpu governor to no avail.

**Comment by jpenala** [comment]
A quick screen recording to illustrate this. The adjustment brush is actually good way to visualise the behaviour from smooth to laggy. You can clearly see the stutter and heavy rubberbanding between cursor and the brush in "laggy mode":

https://streamable.com/hea8d

**Comment by TurboGit** [comment]
When in a lagging mode, is restarting restore a normal speed?

**Comment by jpenala** [comment]
Lagging mode stays the same after restarts. I can only restore normal speed by finding the sweet spot in the panel config. And it stays normal as long as I process images with same aspect ratio. A sweet spot with horizontal/landscape image does not work for vertical/portrait.

**Comment by AlicVB** [comment]
very strange... I'll not have time to test and debug this in the next days, sadly...
At first sight, I would say that : 
1. there is something in pipe or cache we need to reinitialize after panel change, 
2. or we have a module that doesn't like some image size...
3. or something else, not directly related to panel resizing feature

for 1, a restart should have solve it...
for 2, can you test that with a as-minimal-as-possible history (maybe with a jpeg file, to avoid the raw modules) and with different modules ?
for 3, (the most probable cause) : 
- can you find a left-right combo values that ends to lags (call them L and R)
- do you have lags each time left+right values = L+R ?
- if you set left panel value to L+R and set right panel to a value which doesn't lag, closing right panel does ends to lags ?
- revert to commits before the panels resizing changes (3.0, for example) and manually set panels sizes to (L+R)/2 Does it ends to lags ?

Thanks...

**Comment by jpenala** [comment]
Okay I started doing the above debugging and immediately noticed something: 

if the sum of R+L is not dividable by 2, then the darkroom ends up in laggy state. For example if L=250 + R=251 = 501 and  L+R / 2 = 250,5

If I set R=252 and L+R = 502 / 2 = 251 the lag disappears.

Is this just a coincidence and does this make any sense that the panel width division can't end in decimal? I tested this with several values like 300+300 and 300+301, and it always ended up pointing that "dividable by two" does not lag.

**Comment by TurboGit** [comment]
I was suspecting something like that that's why I asked for the width of panels. Thanks a lot for the investigation!

**Comment by jpenala** [comment]
But to contradict my own post above, the lagging values I posted yesterday 

darkroom/ui/0/left_size=318
darkroom/ui/0/right_size=402

are dividable by two, so there must be more to this. Also the fact that some value combinations lag only with certain image proportions...

**Comment by jpenala** [comment]
I've found a magic value that _accepts any image I throw at it_ landscape, portrait, different cameras...

L+R= 648

Any L & R values that have a sum of 648 seem to be okay, and this only applies to a specific setup of minimized filmstrip and hidden rating panel and images zoomed to fit the screen. Any changes to those and the house of card collapses. Zooming and panning the images is laggy but at least the adjustments stay nice throughout. Other screen estate altering variables in this mix are of course the height of the GNOME top bar and height of the adwaita header bar. My screen size btw is  plain regular 1920x1080.

I've extensively tested 3.0.0 with different panel sizes but it stays smooth like it has always been from the 2.7 phase forward.

**Comment by junkyardsparkle** [comment]
> L+R= 648

I tried setting each to 324, but it didn't  fix things for me (1920x1200 display, ~22px  fluxbox taskbar on bottom of screenspace). However, I tried hitting F11 for full screen mode, and this seemed to instantly restore normal operation, even while dragging the panels to various random sizes... I'll try to dig into this a little deeper later, since I have a way to continue using this build now. :-)

**Comment by jpenala** [comment]
> I tried setting each to 324, but it didn't fix things for me (1920x1200 display, ~22px fluxbox taskbar on bottom of screenspace).

The working settings are pretty much tied to the actual screen space darktable gets in the system. I tried this with Fedora Xfce by adjusting the desktop panel size. Every height adjustment threw a previously working dt setup to a laggy mode. Full screen does nothing for me unless I find a specific panel values that work in that mode. 

I have to conclude that everything I've done so far has been inconclusive. Many variables to track, but the laggy vs smooth behaviour is arbitrary as far I know.

**Comment by junkyardsparkle** [comment]
> Full screen does nothing for me unless I find specific panel width values that work in that mode.

Yes, same here, I just got lucky that first time. I also found that changing window size while keeping panel sizes produced the same variation, so I guess it's really a problem with certain scalings of the preview area (possibly relative to image aspect ratio). Still haven't had a chance to look deeper though.

**Comment by AlicVB** [comment]
ok, so I can reproduce too.

After some test, it doesn't seems to be directly related to panels resizing feature : if I compile the branch where I've done the work, everything is fine...
And anyway, with default config, you should have 2x 350 panels, which is the previous default size. And with that sizes, you get lags depending on the top/bottom panels size...

I'll try to find the culprit commit and fix it...

**Comment by AlicVB** [comment]
@jpenala @junkyardsparkle : 
I *think* I've fixed it here : PR #3957 
Can you confirm that it works for you too ? Thanks !

**Comment by junkyardsparkle** [comment]
I can't trigger the issue in that branch... looks good!
